remappers/hostmetrics/filesystem.go

```diff
@@ -0,0 +1,149 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package hostmetrics
+
+import (
+	"strings"
+
+	remappers "github.com/elastic/opentelemetry-lib/remappers/internal"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"golang.org/x/exp/constraints"
+)
+
+func remapFilesystemMetrics(src, out pmetric.MetricSlice,
+	_ pcommon.Resource,
+	dataset string,
+) error {
+	var timestamp pcommon.Timestamp
+	var device, mpoint, fstype string
+	totalUsagePerDevice := make(map[string]int64)
+	totalInodeUsagePerDevice := make(map[string]int64)
+	usedBytesPerDevice := make(map[string]int64)
+
+	for i := 0; i < src.Len(); i++ {
+		metric := src.At(i)
+		switch metric.Name() {
+		case "system.filesystem.usage", "system.filesystem.inodes.usage":
+			dataPoints := metric.Sum().DataPoints()
+			for j := 0; j < dataPoints.Len(); j++ {
+				dp := dataPoints.At(j)
+				value := dp.IntValue()
+				timestamp = dp.Timestamp()
+				deviceValue, mpointValue, fstypeValue, ok := getAttributes(dp)
+				if !ok {
+					continue
+				}
+				device, mpoint, fstype = deviceValue.Str(), mpointValue.Str(), fstypeValue.Str()
+				// Create a unique key for each device
+				deviceKey := device + "_" + mpoint + "_" + fstype
+				if state, ok := dp.Attributes().Get("state"); ok {
+					switch state.Str() {
+					case "used":
+						if metric.Name() == "system.filesystem.usage" {
+							totalUsagePerDevice[deviceKey] += value
+							usedBytesPerDevice[deviceKey] += value
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.used.bytes", device, mpoint, fstype, value)
+						} else {
+							totalInodeUsagePerDevice[deviceKey] += value
+						}
+					case "free":
+						if metric.Name() == "system.filesystem.usage" {
+							totalUsagePerDevice[deviceKey] += value
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.free", device, mpoint, fstype, value)
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.available", device, mpoint, fstype, value)
+						} else {
+							totalInodeUsagePerDevice[deviceKey] += value
+							addFileSystemMetrics(out, timestamp, dataset, "system.filesystem.free_files", device, mpoint, fstype, value)
+						}
+					}
+				}
+			}
+
+		}
+	}
+
+	for deviceKey, totalfsusage := range totalUsagePerDevice {
+		device, mpoint, fstype = parseDeviceKey(deviceKey)
```

Comment:
I am wondering what happens if deviceKey isn't a valid one? The `parseDeviceKey` function should return an error to handle it, and propagate it to the caller, which should then return it to its caller as an error. The `remapFilesystemMetrics` returns an error, but in this it will fail silently.

Reply:
The whole logic of using the device key and then parsing looks like introduced complexity from us. We are first creating the device key by doing a sum and then parsing it(?) I am quite sure we should be able to merge this for loop in the first one but even if we are not we should be able to avoid the re-parsing of the key.

Reply:
Let me try to simplify the logic